drivers/hv/mshv_vtl_main.c

```diff
@@ -801,42 +815,56 @@ void mshv_tdx_request_cache_flush(bool wbnoinvd)
 	__tdx_hypercall(&args);
 }
 
-#define TDCALL_ASM	".byte 0x66,0x0f,0x01,0xcc"
+static void mshv_vtl_on_user_return(struct user_return_notifier *urn)
+{
+	struct mshv_vtl_per_cpu *per_cpu
+		= container_of(urn, struct mshv_vtl_per_cpu, mshv_urn);
+
+	per_cpu->urn_registered = false;
+	user_return_notifier_unregister(urn);
+
+	if (per_cpu->l1_msr_kernel_gs_base != per_cpu->l1_msr_kernel_gs_base_curr) {
```

Comment:
how likely is it we actually hit this branch check where the L2 didn't have a different MSR value than what we expect? i'm wondering, should we just unconditionally restore the previous value if we enter usermode?

Reply:
IE I'm wondering, are we just always failing this check anyways, since we're going to restore the L2's MSRs before we dispatch?

Reply:
I agree in our case we can restore them unconditionally. KVM does the checks but it has more complex conditions to handle e.g. shared OS etc. in which case the MSRs are likely to have not changed.